File: csharp/worldbank-gdp.cs
// To parse this JSON data, add NuGet 'Newtonsoft.Json' then do:
//
//    using QuickType;
//
//    var gdp = Gdp.FromJson(jsonString);

namespace QuickType
{
    using System;
    using System.Collections.Generic;

    using System.Globalization;
    using Newtonsoft.Json;
    using Newtonsoft.Json.Converters;

    public partial class PurpleGdp
    {
        [JsonProperty("indicator")]
        public Country Indicator { get; set; }

        [JsonProperty("country")]
        public Country Country { get; set; }

        [JsonProperty("value")]
        public string Value { get; set; }

        [JsonProperty("decimal")]
        [JsonConverter(typeof(ParseStringConverter))]
        public long Decimal { get; set; }

        [JsonProperty("date")]
        [JsonConverter(typeof(ParseStringConverter))]
        public long Date { get; set; }
    }

    public partial class Country
    {
        [JsonProperty("id")]
        public Id Id { get; set; }

        [JsonProperty("value")]
        public Value Value { get; set; }
    }

    public partial class FluffyGdp
    {
        [JsonProperty("page")]
        public long Page { get; set; }

        [JsonProperty("pages")]
        public long Pages { get; set; }

        [JsonProperty("per_page")]
        [JsonConverter(typeof(ParseStringConverter))]
        public long PerPage { get; set; }

        [JsonProperty("total")]
        public long Total { get; set; }
    }

    public enum Id { Cn, NyGdpMktpCd, Us };

    public enum Value { China, GdpCurrentUs, UnitedStates };

    public partial struct GdpUnion
    {
        public FluffyGdp FluffyGdp;
        public PurpleGdp[] PurpleGdpArray;

        public static implicit operator GdpUnion(FluffyGdp FluffyGdp) => new GdpUnion { FluffyGdp = FluffyGdp };
        public static implicit operator GdpUnion(PurpleGdp[] PurpleGdpArray) => new GdpUnion { PurpleGdpArray = PurpleGdpArray };
    }

    public class Gdp
    {
        public static GdpUnion[] FromJson(string json) => JsonConvert.DeserializeObject<GdpUnion[]>(json, QuickType.Converter.Settings);
    }

    public static class Serialize
    {
        public static string ToJson(this GdpUnion[] self) => JsonConvert.SerializeObject(self, QuickType.Converter.Settings);
    }

    internal static class Converter
    {
        public static readonly JsonSerializerSettings Settings = new JsonSerializerSettings
        {
            MetadataPropertyHandling = MetadataPropertyHandling.Ignore,
            DateParseHandling = DateParseHandling.None,
            Converters = {
                GdpUnionConverter.Singleton,
                IdConverter.Singleton,
                ValueConverter.Singleton,
                new IsoDateTimeConverter { DateTimeStyles = DateTimeStyles.AssumeUniversal }
            },
        };
    }

    internal class GdpUnionConverter : JsonConverter
    {
        public override bool CanConvert(Type t) => t == typeof(GdpUnion) || t == typeof(GdpUnion?);

        public override object ReadJson(JsonReader reader, Type t, object existingValue, JsonSerializer serializer)
        {
            switch (reader.TokenType)
            {
                case JsonToken.StartObject:
                    var objectValue = serializer.Deserialize<FluffyGdp>(reader);
                    return new GdpUnion { FluffyGdp = objectValue };
                case JsonToken.StartArray:
                    var arrayValue = serializer.Deserialize<PurpleGdp[]>(reader);
                    return new GdpUnion { PurpleGdpArray = arrayValue };
            }
            throw new Exception("Cannot unmarshal type GdpUnion");
        }

        public override void WriteJson(JsonWriter writer, object untypedValue, JsonSerializer serializer)
        {
            var value = (GdpUnion)untypedValue;
            if (value.PurpleGdpArray != null)
            {
                serializer.Serialize(writer, value.PurpleGdpArray);
                return;
            }
            if (value.FluffyGdp != null)
            {
                serializer.Serialize(writer, value.FluffyGdp);
                return;
            }
            throw new Exception("Cannot marshal type GdpUnion");
        }

        public static readonly GdpUnionConverter Singleton = new GdpUnionConverter();
    }

    internal class IdConverter : JsonConverter
    {
        public override bool CanConvert(Type t) => t == typeof(Id) || t == typeof(Id?);

        public override object ReadJson(JsonReader reader, Type t, object existingValue, JsonSerializer serializer)
        {
            if (reader.TokenType == JsonToken.Null) return null;
            var value = serializer.Deserialize<string>(reader);
            switch (value)
            {
                case "CN":
                    return Id.Cn;
                case "NY.GDP.MKTP.CD":
                    return Id.NyGdpMktpCd;
                case "US":
                    return Id.Us;
            }
            throw new Exception("Cannot unmarshal type Id");
        }

        public override void WriteJson(JsonWriter writer, object untypedValue, JsonSerializer serializer)
        {
            if (untypedValue == null)
            {
                serializer.Serialize(writer, null);
                return;
            }
            var value = (Id)untypedValue;
            switch (value)
            {
                case Id.Cn:
                    serializer.Serialize(writer, "CN");
                    return;
                case Id.NyGdpMktpCd:
                    serializer.Serialize(writer, "NY.GDP.MKTP.CD");
                    return;
                case Id.Us:
                    serializer.Serialize(writer, "US");
                    return;
            }
            throw new Exception("Cannot marshal type Id");
        }

        public static readonly IdConverter Singleton = new IdConverter();
    }

    internal class ValueConverter : JsonConverter
    {
        public override bool CanConvert(Type t) => t == typeof(Value) || t == typeof(Value?);

        public override object ReadJson(JsonReader reader, Type t, object existingValue, JsonSerializer serializer)
        {
            if (reader.TokenType == JsonToken.Null) return null;
            var value = serializer.Deserialize<string>(reader);
            switch (value)
            {
                case "China":
                    return Value.China;
                case "GDP (current US$)":
                    return Value.GdpCurrentUs;
                case "United States":
                    return Value.UnitedStates;
            }
            throw new Exception("Cannot unmarshal type Value");
        }

        public override void WriteJson(JsonWriter writer, object untypedValue, JsonSerializer serializer)
        {
            if (untypedValue == null)
            {
                serializer.Serialize(writer, null);
                return;
            }
            var value = (Value)untypedValue;
            switch (value)
            {
                case Value.China:
                    serializer.Serialize(writer, "China");
                    return;
                case Value.GdpCurrentUs:
                    serializer.Serialize(writer, "GDP (current US$)");
                    return;
                case Value.UnitedStates:
                    serializer.Serialize(writer, "United States");
                    return;
            }
            throw new Exception("Cannot marshal type Value");
        }

        public static readonly ValueConverter Singleton = new ValueConverter();
    }

    internal class ParseStringConverter : JsonConverter
    {
        public override bool CanConvert(Type t) => t == typeof(long) || t == typeof(long?);

        public override object ReadJson(JsonReader reader, Type t, object existingValue, JsonSerializer serializer)
        {
            if (reader.TokenType == JsonToken.Null) return null;
            var value = serializer.Deserialize<string>(reader);
            long l;
            if (Int64.TryParse(value, out l))
            {
                return l;
            }
            throw new Exception("Cannot unmarshal type long");
        }

        public override void WriteJson(JsonWriter writer, object untypedValue, JsonSerializer serializer)
        {
            if (untypedValue == null)
            {
                serializer.Serialize(writer, null);
                return;
            }
            var value = (long)untypedValue;
            serializer.Serialize(writer, value.ToString());
            return;
        }

        public static readonly ParseStringConverter Singleton = new ParseStringConverter();
    }
}

File: golang/worldbank-gdp.go
// To parse and unparse this JSON data, add this code to your project and do:
//
//    gdp, err := UnmarshalGdp(bytes)
//    bytes, err = gdp.Marshal()

package main

import "bytes"
import "errors"
import "encoding/json"

type Gdp []GDPUnion

func UnmarshalGdp(data []byte) (Gdp, error) {
	var r Gdp
	err := json.Unmarshal(data, &r)
	return r, err
}

func (r *Gdp) Marshal() ([]byte, error) {
	return json.Marshal(r)
}

type PurpleGDP struct {
	Indicator Country `json:"indicator"`
	Country   Country `json:"country"`  
	Value     *string `json:"value"`    
	Decimal   Decimal `json:"decimal"`  
	Date      string  `json:"date"`     
}

type Country struct {
	ID    ID    `json:"id"`   
	Value Value `json:"value"`
}

type FluffyGDP struct {
	Page    int64  `json:"page"`    
	Pages   int64  `json:"pages"`   
	PerPage string `json:"per_page"`
	Total   int64  `json:"total"`   
}

type ID string
const (
	CN ID = "CN"
	NyGdpMktpCD ID = "NY.GDP.MKTP.CD"
	Us ID = "US"
)

type Value string
const (
	China Value = "China"
	GDPCurrentUS Value = "GDP (current US$)"
	UnitedStates Value = "United States"
)

type Decimal string
const (
	The0 Decimal = "0"
)

type GDPUnion struct {
	FluffyGDP      *FluffyGDP
	PurpleGDPArray []PurpleGDP
}

func (x *GDPUnion) UnmarshalJSON(data []byte) error {
	x.PurpleGDPArray = nil
	x.FluffyGDP = nil
	var c FluffyGDP
	object, err := unmarshalUnion(data, nil, nil, nil, nil, true, &x.PurpleGDPArray, true, &c, false, nil, false, nil, false)
	if err != nil {
		return err
	}
	if object {
		x.FluffyGDP = &c
	}
	return nil
}

func (x *GDPUnion) MarshalJSON() ([]byte, error) {
	return marshalUnion(nil, nil, nil, nil, x.PurpleGDPArray != nil, x.PurpleGDPArray, x.FluffyGDP != nil, x.FluffyGDP, false, nil, false, nil, false)
}

func unmarshalUnion(data []byte, pi **int64, pf **float64, pb **bool, ps **string, haveArray bool, pa interface{}, haveObject bool, pc interface{}, haveMap bool, pm interface{}, haveEnum bool, pe interface{}, nullable bool) (bool, error) {
	if pi != nil {
		*pi = nil
	}
	if pf != nil {
		*pf = nil
	}
	if pb != nil {
		*pb = nil
	}
	if ps != nil {
		*ps = nil
	}

	dec := json.NewDecoder(bytes.NewReader(data))
	dec.UseNumber()
	tok, err := dec.Token()
	if err != nil {
		return false, err
	}

	switch v := tok.(type) {
	case json.Number:
		if pi != nil {
			i, err := v.Int64()
			if err == nil {
				*pi = &i
				return false, nil
			}
		}
		if pf != nil {
			f, err := v.Float64()
			if err == nil {
				*pf = &f
				return false, nil
			}
			return false, errors.New("Unparsable number")
		}
		return false, errors.New("Union does not contain number")
	case float64:
		return false, errors.New("Decoder should not return float64")
	case bool:
		if pb != nil {
			*pb = &v
			return false, nil
		}
		return false, errors.New("Union does not contain bool")
	case string:
		if haveEnum {
			return false, json.Unmarshal(data, pe)
		}
		if ps != nil {
			*ps = &v
			return false, nil
		}
		return false, errors.New("Union does not contain string")
	case nil:
		if nullable {
			return false, nil
		}
		return false, errors.New("Union does not contain null")
	case json.Delim:
		if v == '{' {
			if haveObject {
				return true, json.Unmarshal(data, pc)
			}
			if haveMap {
				return false, json.Unmarshal(data, pm)
			}
			return false, errors.New("Union does not contain object")
		}
		if v == '[' {
			if haveArray {
				return false, json.Unmarshal(data, pa)
			}
			return false, errors.New("Union does not contain array")
		}
		return false, errors.New("Cannot handle delimiter")
	}
	return false, errors.New("Cannot unmarshal union")

}

func marshalUnion(pi *int64, pf *float64, pb *bool, ps *string, haveArray bool, pa interface{}, haveObject bool, pc interface{}, haveMap bool, pm interface{}, haveEnum bool, pe interface{}, nullable bool) ([]byte, error) {
	if pi != nil {
		return json.Marshal(*pi)
	}
	if pf != nil {
		return json.Marshal(*pf)
	}
	if pb != nil {
		return json.Marshal(*pb)
	}
	if ps != nil {
		return json.Marshal(*ps)
	}
	if haveArray {
		return json.Marshal(pa)
	}
	if haveObject {
		return json.Marshal(pc)
	}
	if haveMap {
		return json.Marshal(pm)
	}
	if haveEnum {
		return json.Marshal(pe)
	}
	if nullable {
		return json.Marshal(nil)
	}
	return nil, errors.New("Union must not be null")
}
